### DISCUSSION TITLE
Autodiscover Filter: re2 fails to parse regex with negative lookahead assertion

### DISCUSSION THREAD

**Discussion by Stefan-J-Hanke** [renovatebot/renovate#16472] [Request Help]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

32.105.4

### Please select which platform you are using if self-hosting.

GitLab self-hosted

### If you're self-hosting Renovate, tell us what version of the platform you run.

15.0

### Was this something which used to work for you, and then stopped?

I never saw this working

### Describe the bug

I'd like to autodiscover every repository below a certain GitLab Group _except_ a single one.

Fortunately, Renovate not only allows to use minimatch, but regular expressions also (https://docs.renovatebot.com/self-hosted-configuration/#autodiscoverfilter), so it should be possible using a negative lookahead assertion (from an SO answer in another context)

The autodiscover filter in question is  **"/groupname1/groupname2/(?!excluded-project-name).*/"** which matches every direct project below **groupname1/groupname2** except **excluded-project-name**

I noticed that renovate is preferring to use "re2" as regex library instead of the node one. "re2" just fails to parse this regular expression - the default one can parse it. I've forcefully removed "re2" (by just "rm -rf"ing the folder **/usr/src/app/node_modules/re2**) to verify this, however that's not a line of code that can stay.

### Relevant debug logs

<details><summary>Logs</summary>

```
"loggerErrors": [
         {
           "name": "renovate",
           "level": 60,
           "logContext": "_z-kTL3a9lsDo6wDD_aPj",
           "err": {
             "message": "Failed to parse regex pattern \"/groupname1/groupname2/(?!excluded-project-name).*/\"",
             "stack": "Error: Failed to parse regex pattern \"/groupname1/groupname2/(?!excluded-project-name).*/\"\n    at autodiscoverRepositories (/usr/src/app/node_modules/renovate/lib/workers/global/autodiscover.ts:37:15)\n    at processTicksAndRejections (internal/process/task_queues.js:95:5)\n    at Object.start (/usr/src/app/node_modules/renovate/lib/workers/global/index.ts:131:14)\n    at /usr/src/app/node_modules/renovate/lib/renovate.ts:16:22"
           },
           "msg": "Fatal error: Failed to parse regex pattern \"/groupname1/groupname2/(?!excluded-project-name).*/\""
         }
]
```

</details>

### Have you created a minimal reproduction repository?

No reproduction repository

**Comment by secustor** [comment]
This is expected behaviour as RE2 does not support look arounds. 

> It is also not a goal to implement all of the features offered by Perl, PCRE and other engines. As a matter of principle, RE2 does not support constructs for which only backtracking solutions are known to exist. Thus, backreferences and look-around assertions are not supported.

https://github.com/google/re2/wiki/WhyRE2

What you can do is, to use an internal feature regex negation   
`!/mygroup/example-project/`

https://docs.renovatebot.com/self-hosted-configuration/#autodiscoverfilter

**Reply by fgreinacher** [reply]
How would this look like for the example? I think you need to chose between taking the regex as-is or negating it completely.

**Reply by secustor** [reply]
I would give the Renovate user only access to this Gitlab group and then negate using the regex above.

**Reply by fgreinacher** [reply]
Hm, that works in some scenarios, but not generally. I'd rather not have an autodiscover filter that makes Renovate look at each and every project it might have access to.

**Reply by secustor** [reply]
I have started to work again on https://github.com/renovatebot/renovate/issues/8763 so this can be helpful as well as the somewhere discussed `autodiscoverExcludeFilter` option.

**Reply by Stefan-J-Hanke** [reply]
I didn't fully get what actually will change and how I would specify that negative lookahead regex, esp. considering that filter array will be ORed together.

**Comment by fgreinacher** [comment]
Would it be an option to opt-out of RE2 for self-hosted?

**Reply by rarkins** [reply]
It's an optional dependency so you might be able to simply delete it out of `node_modules`. You'll get a warning at startup but otherwise it will load.

**Reply by fgreinacher** [reply]
Yes, thats what @Stefan-J-Hanke did, but it feels a little bit hardcore 😬
